crates/polars-core/src/chunked_array/logical/struct_/mod.rs

```diff
@@ -173,58 +177,54 @@ impl StructChunked {
             field,
             chunks: vec![arrow_array],
             null_count: 0,
+            total_null_count: 0,
         };
         out.set_null_count();
         out
     }
 
     fn set_null_count(&mut self) {
-        let mut null_count = 0;
-        let chunks_lens = self.fields()[0].chunks().len();
-
-        // fast path
-        // we early return if a column doesn't have nulls
-        for i in 0..chunks_lens {
-            for s in self.fields() {
-                let arr = &s.chunks()[i];
-                let has_nulls = arr.null_count() > 0 || matches!(s.dtype(), DataType::Null);
-                if !has_nulls {
-                    self.null_count = 0;
-                    return;
-                }
-            }
+        // Count both the total number of nulls and the rows where everything is null
+        (self.null_count, self.total_null_count) = (0, 0);
+
+        // If there is at least one field with no null values, no rows are null. However, we still
+        // have to count the number of nulls per field to get the total number. Fortunately this is
+        // cheap since null counts per chunk are pre-computed.
+        let (could_have_nulls, total_null_count) =
+            self.fields().iter().fold((true, 0), |acc, s| {
+                (acc.0 & (s.null_count() != 0), acc.1 + s.null_count())
+            });
+        self.total_null_count = total_null_count;
+        if !could_have_nulls {
```

Comment:
Why do we have an extra `could_have_nulls` variable? Isn't that the same as `total_null_count != 0`?

As I understand it, if `total_null_count == 0`, we can return early, otherwise we have to check the rows.

Or do I misunderstand it here?

Reply:
You can have nulls only if the count in every field is > 0. So in addition to `total_null_count == 0` you can return early when any one field has no nulls. For example consider a struct with three fields

|f1|f2|f3|
|-|-|-|
|1|null|null|
|2|'a'|null|
|3|null|'b'|

The total null count is 4, but we can still return early because the first field has no nulls.

Reply:
Ah, right. This is confusing as the `null` name now collides with two meanings.

Shall we name one `null_row` and the other `null_value` to disambiguate and reflect that in the parameter names. That should make it clear which one we mean.

Reply:
You mean instead of `null_count` and `total_null_count`? Or just here to indicate the potential for a null row and separate out the count of all null values across all fields?

Reply:
I renamed `could_have_nulls` to make it clearer we're talking about entire rows. Hopefully that's what you meant. I think that, combined with the comment about should make the distinction clear. Since after that section we're done with the total number of nulls across all fields.